[WIP] Create auto-generate site with markdown-docs

ericstoekl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR is a continuation of the work started by @johnmccabe on the `markdown-docs` site (#274). John did the bulk of this work and it is fantastic so far, but we need to figure out what else to include/exclude from the documentation site.

## Description
<!--- Describe your changes in detail -->
Check out this PR, and use `mkdocs build` to create the site from the `docs/` folder. User`mkdocs gh-deploy` to deploy the site to your github.io account. You can check out my sample at https://ericstoekl.github.io/faas/

Changes since John's PR:

- Rebase PR to latest version of `master` branch, so it is ready to merge
- Fixup some typos and errors in the _Deployment_ section of the docs
- Add the `CONTRIBUTING.md` file to the docs site.

This will automatically be built with Travis-CI if desired.

### Next Steps

We need suggestions on how to make this documentation site better and more useful. I think it looks good in its current state, but we could use suggestions on how to make it better.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md)) (#253)


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

JockDaRock commented:
Organizing docs is not really favorite, lol, but we need to talk about what a good docs site should look like for OpenFaaS. This would be beyond asethetics of the UI and just including the current guides. At least I my opinion anyway. 🤔

I think this is a big beast to tackle in a lot of ways and may be beyond the scope of just getting a docs site up, but it is something we need to think about and take a piece at a time. 

Jock's thoughts on docs ☺️.

BurtonR commented:
I agree with @JockDaRock that the docs site/page should be clean and well organized, but I also think that we need to get _something_ out there sooner than later. A central place for all the documentation is better than the current situation of have a readme in several different places spread across several repos.

I think we should push to get this out there, then as it gets used more, we can start to see where it can be improved.

alexellis commented:
I'm not sure I understand the point @JockDaRock is making here, but the docs PR LGTM. I just want to make sure we have it in sync with all the changes to guides / the guide index and any points extracted from the README files. Polishing can be done in-flight.

alexellis commented on .travis.yml:
```diff
@@ -1,33 +1,15 @@
-addons:
-  apt:
-    packages:
-      - docker-ce
```
Why is this deleted?

alexellis commented on .travis.yml:
```diff
@@ -1,33 +1,15 @@
+  - docker run --rm -it -v $PWD:/docs squidfunk/mkdocs-material build --verbose --clean --strict
+
 deploy:
-  provider: releases
```
These elements cannot be deleted?

ericstoekl replied:
I believe that these deletions were a bungled merge/merge conflict resolution. I'll put these changes back

alexellis commented on DEV.md:
```diff
@@ -1,95 +0,0 @@
-## Develop your own function
```
Why was this file deleted?

ericstoekl replied:
I'll put this back

ericstoekl commented:
Updated to resolve merge conflict and re-add the accidental deletions of DEV.md and some data from .travis.yml

ericstoekl commented:
Update to include Tutorials and Troubleshooting. Check it out: https://ericstoekl.github.io/faas/

ericstoekl commented:
Refresh to resolve conflicts. Site is updated at location https://ericstoekl.github.io/faas/

ericstoekl commented:
Updated this PR again, resolving conflicts and syncing up the [Deployment for Kubernetes guide](https://ericstoekl.github.io/faas/deployment/kubernetes/)

BurtonR commented:
Just some minor link issues. Otherwise it looks great! 

* When viewing "Part 2 - OpenFaaS CLI" the title and right menu indicate that "Part 4" is being viewed. Both "Part 2" and "Part 4" are highlighted on the nav menu
* Clicking "Part 4 - Create a Function" on the nav menu takes you to the same page as "Page 2"
* Also, "Part 2" is more "Test Drive" than CLI. 
* Clicking the "Begin the Test Drive" link just points to the same doc in the repo
* The "Invoke the sample function with curl or Postman" header only have text to star the repo...
* The link at the bottom of the "Docker Swarm" page: "Step 2 - OpenFaaS UIs" links back to the Docker Swarm page. Should just be removed since there really isn't a "step 2"

On another note, maybe we could clean up the docs to avoid repeating ourselves. A lot of the pages talk about setting up, or deploying, OpenFaaS. It makes sense in the repo readmes since they're disconnected, but with the docs site, it should either be assumed that the reader is already set up, or just add a message like "This guide assumes OpenFaaS is deployed locally" with a link to the deployment guide page.

ericstoekl commented:
Thanks @BurtonR . I'm going to need to refactor the [TestDrive page](https://github.com/openfaas/faas/pull/464/files#diff-928c7848b554a3d9bd537aefd82581e3). It needs to be a step by step breakdown of the features of faas-cli:

1. New
2. Build
3. Push
4. Deploy
5. Store

And I think there are other places where the docs would benefit from deduplication

kenfdev commented:
I'm also checking if we aren't missing any topics spread across the repos. Should be able to share this week. Basically I have the same opinion as @BurtonR but I'm also thinking the target audience of the `Development` is obscure. Meaning, is it for OpenFaaS contributor? or is it for people who are actually going to develop with OpenFaaS.

kenfdev commented:
I've roughly checked if we're missing some resource across the repositories. Things at least I found was

* [Deployment guide for DigitalOcean](https://github.com/openfaas/faas/blob/master/guide/deployment_digitalocean.md#deployment-guide-for-digitalocean)
* [OpenFaaS Function Store](https://github.com/openfaas/store)

As have been mentioned, we should have a CLI chapter, too.
By **syncing** I think the ones above are missing.

About the docs' content itself, I think:

* The portability of OpenFaaS isn't shown well. It would be better if we could say "While deploying OpenFaaS itself may differ between orchestrators(Swarm, Kubernetes, etc.), after that, everything is the same. So the [Deployment Guides](https://ericstoekl.github.io/faas/deployment/swarm/) ' next step should all go the same `Part 2` like page.
* [Sample Functions](https://ericstoekl.github.io/faas/getting-started/sample_functions/) could be improved, or maybe we don't need this page
* We might not need the Roadmap and Contributing page for the docs. Github may be enough.
* [API Gateway](https://ericstoekl.github.io/faas/developer/gateway/) and [Dashboard UI](https://ericstoekl.github.io/faas/developer/dashboard/) are pretty much identical
* [Secret Management](https://ericstoekl.github.io/faas/operations/secure_secret_management/) should have a kubernetes version, too.

This is more than just collecting the markdowns across the repositories. But it's what I thought as I was checking the site.

ericstoekl commented:
Update and rebase to latest version of `master` branch. Cleaned out some unneeded content from the CLI section. Removed the files mentioned by Ken, also added "Cloud Platform" specific guides section. Still remaining tasks to do:

- Improve CLI section (so far have only shortened it some)
- Add OpenFaaS Store section
- Part 3 (Create a Function) could be improved quite a bit
- Clean up "Development" section, figure out its purpose.

alexellis commented:
Hi @ericstoekl we have some new merge conflicts here. Let's get on a call in the next 2 weeks to discuss how to merge this and take it forward before it gets out of sync again.

alexellis commented:
Update from contributors' call:
* The docs PR is getting out of sync really fast and is hard to maintain
* The docs cover too much - i.e. the README files appear to be built into the docs
* Having the docs in the faas repo means cutting a new set of watchdog binaries and new gateway image for every docs / typo change.. we need to come up with a better approach
* Existing guides / content needs to redirect to the appropriate section / content in the new docs
* A server has to be provisioned with TLS - this takes dedicated time / effort and maintenance

@justincormack mentioned something being used by Docker to host/generate the docs I should look that that and it would probably be able to use the existing mechanism.

On the positive side:

* We have a nice theme designed by @johnmccabe 
* It builds quite fast and can be processed on Travis
* A couple of people have been involved and liked the structure, so we have something to build upon

justincormack commented:
It was Netlify....


On 9 Mar 2018 23:11, "Alex Ellis" <notifications@github.com> wrote:

> Update from contributors' call:
>
>    - The docs PR is getting out of sync really fast and is hard to
>    maintain
>    - The docs cover too much - i.e. the README files appear to be built
>    into the docs
>    - Having the docs in the faas repo means cutting a new set of watchdog
>    binaries and new gateway image for every docs / typo change.. we need to
>    come up with a better approach
>    - Existing guides / content needs to redirect to the appropriate
>    section / content in the new docs
>    - A server has to be provisioned with TLS - this takes dedicated time
>    / effort and maintenance
>
> @justincormack <https://github.com/justincormack> mentioned something
> being used by Docker to host/generate the docs I should look that that and
> it would probably be able to use the existing mechanism.
>
> On the positive side:
>
>    - We have a nice theme designed by @johnmccabe
>    <https://github.com/johnmccabe>
>    - It builds quite fast and can be processed on Travis
>    - A couple of people have been involved and liked the structure, so we
>    have something to build upon
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openfaas/faas/pull/464#issuecomment-371971680>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdcPGo8Qj1coC269PrTCokIjVtBQ934ks5tcwwOgaJpZM4RPe8X>
> .
>

alexellis commented:
Closing this pull request as docs.openfaas.com is live now and we have GitHub.com/openfaas/docs for the markdown. We'll continue to take parts of this PR over as it makes sense for new pages.

alexellis commented:
@justincormack Netlify works great thank you 👍
